src/Tools/ExternalAccess/Extensions/External/IExtensionWorkspaceMessageHandler.cs

```diff
@@ -0,0 +1,38 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.Extensions;
+
+/// <summary>
+/// The base interface required to implement an extension message handler for non-document-specific messages.
+/// </summary>
+/// <typeparam name="TMessage">The type of the message received by the hander. <typeparamref name="TMessage"/> must be
+/// serializable using System.Text.Json.</typeparam>
+/// <typeparam name="TResponse">The type of the response returned by the hander. <typeparamref name="TResponse"/> must
+/// be serializable using System.Text.Json.</typeparam>
+/// <remarks>
+/// <para>
+/// Extension message handlers allow IDE extensions to send messages to the compiler and receive responses. This is
+/// useful for scenarios where the IDE needs to query the compilation state or other information from the compiler.
+/// </para>
+/// <para>
+/// The lifetime of a message handler object is tied to the solution: when the solution is closed, the assemblies that
```

Comment:
this needs to be updated to mention we may create multiple instances (but only 1 wins)